### PR TITLE
Patch presburger parser

### PR DESCRIPTION
**Pull request by Dinistro**
Currently, I have some open questions/problems: 
- We have some overlapping between Presburger and Affine parser
   I do not know if there is a reasonable way to eliminate these.
- Decide where to place the parser, i.e. inside the `Analysis/Presburger` folder or inside the `Parser/` folder?
- Where to integrate it into CMake, it should not be built as part of mlir-opt as of yet. 
Currently it's part of MLIRPresburger
- Affine constraints enforce a zero on the right hand side, should we do the same? 
- Should we check for EOF? 
This is currently done
- How to deal with the Coefficients? The struct solution is not that nice, IMHO
- What about symbols? They are not checked in arjun’s unit-test.

There is still some work to do, as listed here:

- Refactoring
    - The Coefficient handling needs some love, especially the merging part. 
- More unit-tests?

### CODE REVIEW

File: mlir/lib/Analysis/CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,9 @@ add_mlir_library(MLIRLoopAnalysis
   LoopAnalysis.cpp
   NestedMatcher.cpp
   PresburgerSet.cpp
+  Presburger/Simplex.cpp
+  Presburger/Matrix.cpp
+  Presburger/Parser.cpp
```
Comment (superty) [inline mlir/lib/Analysis/CMakeLists.txt]: I don't think this is needed since it already links with MLIRPresburger.

File: mlir/lib/Parser/DialectSymbolParser.cpp [hunk 2]
```diff
@@ -81,7 +81,7 @@ class CustomDialectAsmParser : public DialectAsmParser {
   }
 
   /// Parse an optional integer value from the stream.
-  OptionalParseResult parseOptionalInteger(uint64_t &result) override {
+  OptionalParseResult parseOptionalInteger(int64_t &result) override {
```
Comment (superty) [inline mlir/lib/Parser/DialectSymbolParser.cpp]: Are you sure it's fine to change these (and below)? What about creating an additional `parseOptimalSignedInteger`?

File: mlir/unittests/Analysis/Presburger/CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,10 @@
 add_mlir_unittest(MLIRPresburgerTests
   MatrixTest.cpp
   SimplexTest.cpp
+  ParserTest.cpp
 )
 
 target_link_libraries(MLIRPresburgerTests
-  PRIVATE MLIRPresburger)
+  PRIVATE MLIRPresburger
+  MLIRLoopAnalysis)
```
Comment (superty) [inline mlir/unittests/Analysis/Presburger/CMakeLists.txt]: This creates a cyclic dependency since MLIRLoopAnalysis depends on MLIRPresburger... (why is this required?)

File: mlir/include/mlir/Analysis/Presburger/Parser.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains the interface to the Presburger parser.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef PRESBURGER_PARSER_H
+#define PRESBURGER_PARSER_H
```
Comment (superty) [inline mlir/include/mlir/Analysis/Presburger/Parser.h]: `MLIR_ANALYSIS_PRESBURGER_PARSER_H`

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a parser for Presburger sets.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/Presburger/Parser.h"
+#include "../../Parser/Parser.h"
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: I don't think we should be doing this. Maybe they only expect people to use these headers from inside the `Parser` directory?

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+// TODO: renaming and comments
+struct Coefficients {
+  int64_t constant;
+  SmallVector<int64_t, 8> coeffs;
+};
+
+enum class ConstraintKind { LE, GE, EQ };
+
+/// This is a specialized parser for Presburger structures (Presburger set)
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: Please put a full stop at the end of sentences in comments.

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+      kind = ConstraintKind::LE;
+      break;
+    }
+    return emitError("strict inequalities are not supported");
+  case Token::bang:
+    consumeToken(Token::bang);
+    if (consumeIf(Token::equal)) {
+      return emitError("!= constraints are not supported");
+    }
+    [[clang::fallthrough]];
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: Why is "expected comparison operator" emitted for `!=` but not `<`, `>`?

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+                                          Token::Kind rightToken) {
+  auto parseElt = [&]() -> ParseResult {
+    StringRef name;
+    if (parseVariable(name))
+      return failure();
+
+    auto it = dimAndSymNameToIndex.find(name);
+    if (it != dimAndSymNameToIndex.end())
+      return emitError(
+          "repeated variable names in the tuple are not yet supported");
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: It's unclear whether we will support this later. It probably makes sense to just say that names shouldn't repeat

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+using llvm::SMLoc;
+using llvm::SourceMgr;
+using llvm::StringMap;
+
+namespace {
+
+// TODO: renaming and comments
+struct Coefficients {
+  int64_t constant;
+  SmallVector<int64_t, 8> coeffs;
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: Why not just use the same format as in FAC -- last element of `coeffs` is the constant term.

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+  bool intFound = false;
+  bool varFound = false;
+
+  if (parseOptionalInteger(coeff).hasValue()) {
+    intFound = true;
+    if (isNegated)
+      coeff = -coeff;
+  } else if (isNegated) {
+    coeff = -1;
+    intFound = true;
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: It seems like this would cause a `-` followed by nothing to be accepted as a constant term `-1`.

Comment (superty) [reply mlir/lib/Analysis/Presburger/Parser.cpp]: You can just take the isNegated part out IMO
```cpp
  if (parseOptionalInteger(coeff).hasValue())
    intFound = true;
  if (isNegated)
    coeff = -coeff;
```

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  if (getToken().is(Token::bare_identifier)) {
+    if (parseVariable(identifier))
+      return failure();
+    varFound = true;
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: I think the following would be simpler

```cpp
  if (getToken().isNot(Token::bare_identifier)) {
    if (!intFound)
      return emitError("expected non-empty term");
    // add to constant
    return success();
  }
  
  // handle identifier case
}
```

Which allows you to get rid of `varFound` entirely.

File: mlir/lib/Analysis/Presburger/Parser.cpp [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+/// Parse a PresburgerSet from a given StringRef
+FailureOr<PresburgerSet> mlir::parsePresburgerSet(StringRef str,
+                                                  MLIRContext *ctx) {
+  SourceMgr sourceMgr;
+  sourceMgr.AddNewSourceBuffer(MemoryBuffer::getMemBuffer(str), SMLoc());
+
+  SymbolState symbols;
+  ParserState state(sourceMgr, ctx, symbols);
+  PresburgerParser parser(state);
+  PresburgerSet set = PresburgerSet::getUniverse(1, 1);
```
Comment (superty) [inline mlir/lib/Analysis/Presburger/Parser.cpp]: I would add a comment mentioning that the value of `set` at this point doesn't matter since it gets overwritten by the parser.

File: mlir/unittests/Analysis/Presburger/ParserTest.cpp [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+                                            {-1, 0, 10},  // x <= 10.
+                                            {0, -1, 10},  // y <= 10.
+                                            {1, 1, -2},   // x + y >= 2.
+                                            {-1, -1, 30}, // x + y <= 30.
+                                            {1, -1, 0},   // x - y >= 0.
+                                            {-1, 1, 10},  // x - y <= 10.
+                                        })});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
```
Comment (superty) [inline mlir/unittests/Analysis/Presburger/ParserTest.cpp]: Some more possible test cases

`(x) : ` (invalid)
`(x) : (x >= 0 and x <= 5) or (x >= -2 and x <= -5)`
`(x) : (x - >= -2)` (invalid)
`(x) : (1x >= -)` (invalid)
`(x) : (-x - -2 >= - -2)` (invalid)
`(x) : (-1x + -2 >= +2)`
`(x) : (1x -2 >= -2)`

**Review by superty — COMMENTED** [review]
I think it's better to put it into `Parser/`, it doesn't seem correct to include `../../Parser.h` from Presburger.

I think we should use the same form as FlatAffineConstraints, so just a vector of coefficients where the last element is a constant. This also means you don't need the struct if you don't want to use it.

I think it makes sense to check for EOF, IIUC `(x) : (x >= 0)))))` would be considered a valid set otherwise. The full input to the parser should form a valid set.

I think it's ok to ignore symbols for the first patch.

File: mlir/lib/Analysis/Presburger/Parser.cpp (hunk 7 above)
Comment (Dinistro) [reply mlir/lib/Analysis/Presburger/Parser.cpp]: I do not understand what you mean. 
We emit "!= constraints are not supported" and "strict (in)equalities are not supported". The fallthrough only happens if it's only `!`.

**Comment by superty** [comment]
Ah I see, my bad.

File: mlir/lib/Parser/DialectSymbolParser.cpp (hunk 2 above)
Comment (Dinistro) [reply mlir/lib/Parser/DialectSymbolParser.cpp]: There is some inconsistency in Parser::parseOptionalInteger (see https://github.com/Superty/llvm-project/blob/patch-presburger-parser/mlir/lib/Parser/Parser.cpp#L98). They actually parse a signed integer (first the sign and then a UInt64) and then return it as a int64_t. Unfortunately, this result is transformed to a `uint64_t` in this method here. 

I figured that this is a bug and decided to change it.

Comment (Dinistro) [reply mlir/lib/Parser/DialectSymbolParser.cpp]: Not sure what I should do here. Maybe keep it and discuss this in the upstream review.

File: mlir/lib/Analysis/Presburger/Parser.cpp (hunk 5 above)
Comment (Dinistro) [reply mlir/lib/Analysis/Presburger/Parser.cpp]: I assumed the same thing. The problem is, that this parser is only used for Presburger unit-tests for now.

File: mlir/unittests/Analysis/Presburger/CMakeLists.txt (hunk 3 above)
Comment (Dinistro) [reply mlir/unittests/Analysis/Presburger/CMakeLists.txt]: The `PresburgerSet` is required and it isn't part of `MLIRPresburger`. The cyclic dependency isn't a problem, AFAIK, at least not on my machine.

Comment (Dinistro) [reply mlir/unittests/Analysis/Presburger/CMakeLists.txt]: Any idea on how to resolve this?

File: mlir/lib/Parser/DialectSymbolParser.cpp (hunk 2 above)
Comment (superty) [reply mlir/lib/Parser/DialectSymbolParser.cpp]: That sounds good to me.

File: mlir/unittests/Analysis/Presburger/CMakeLists.txt (hunk 3 above)
Comment (superty) [reply mlir/unittests/Analysis/Presburger/CMakeLists.txt]: You will have to move the parser out of `Analysis/Presburger`. Either put it in `Analysis` with `PresburgerSet` or move it into `Parser`.

Comment (superty) [inline mlir/unittests/Analysis/Presburger/CMakeLists.txt]: Wait, these are the dependencies for MLIRPresburgerTests, not MLIRPresburger. My bad, there is no cyclic dependency here then

File: mlir/unittests/Analysis/Presburger/ParserTest.cpp (hunk 13 above)
Comment (Dinistro) [reply mlir/unittests/Analysis/Presburger/ParserTest.cpp]: Currently, the set `(x) : (-1x + -2 >= +2)` cannot be parsed, as Parser::parseOptionalInteger does not allow a leading `+`.

Comment (superty) [reply mlir/unittests/Analysis/Presburger/ParserTest.cpp]: I see, okay.

**Comment by Dinistro** [comment]
I just found out that the parseOptionalInteger is even more broken than expected. It will consume a `Token::Minus`, even if the integer parsing fails afterwards. I probably have to do something ugly to get around this issue :(

**Comment by Dinistro** [comment]
I'm closing this due to it being outdated and because something with the amount of commits seems odd.
